Add multi-sig example binaries

v9ai opened the pull request:
Related to: https://github.com/hyperliquid-dex/hyperliquid-rust-sdk/issues/143
- convert_to_multi_sig_user: Convert user to multi-sig with authorized users and threshold
- multi_sig_order: Execute multi-sig orders with JSON and typed actions
- multi_sig_register_token: Multi-sig token registration with spot deploy
- multi_sig_usd_send: Multi-sig USD transfers with signature chain parameters

Commit:
Add multi-sig example binaries

- convert_to_multi_sig_user: Convert user to multi-sig with authorized users and threshold
- multi_sig_order: Execute multi-sig orders with JSON and typed actions
- multi_sig_register_token: Multi-sig token registration with spot deploy
- multi_sig_usd_send: Multi-sig USD transfers with signature chain parameters

Examples show structure for future implementation when multi-sig functionality is added to Rust SDK

nicolad commented:
@lmlmt note that multi-sig conversion isn’t implemented in the Rust SDK yet; the examples in this PR are scaffolding to show the shape/params. Once this PR is merged, I can look into adding the actual conversion support (or exposing a minimal API surface) and follow up with a dedicated PR

lmlmt commented:
> @lmlmt note that multi-sig conversion isn’t implemented in the Rust SDK yet; the examples in this PR are scaffolding to show the shape/params. Once this PR is merged, I can look into adding the actual conversion support (or exposing a minimal API surface) and follow up with a dedicated PR

@nicolad prefer having the multi-sig functionality implemented as well, for completeness

v9ai commented:
> > @lmlmt note that multi-sig conversion isn’t implemented in the Rust SDK yet; the examples in this PR are scaffolding to show the shape/params. Once this PR is merged, I can look into adding the actual conversion support (or exposing a minimal API surface) and follow up with a dedicated PR
> 
> @nicolad prefer having the multi-sig functionality implemented as well, for completeness

thank you @lmlmt 
will add today

Commit:
Add complete multi-signature functionality with EIP-712 signing, account conversion, address management, and transaction operations for orders, USDC transfers, and spot transfers including comprehensive tests and examples.

luca992 commented on src/exchange/exchange_client.rs:
```diff
@@ -867,6 +905,176 @@ impl ExchangeClient {
 
         self.post(action, signature, timestamp).await
     }
+
+    // Multi-sig methods
+
+    async fn post_multi_sig(
+        &self,
+        action: serde_json::Value,
+        signatures: Vec<Signature>,
+        nonce: u64,
+    ) -> Result<ExchangeResponseStatus> {
+        let exchange_payload = MultiSigExchangePayload {
+            action,
+            signatures,
+            nonce,
+            vault_address: self.vault_address,
+        };
+        let res = serde_json::to_string(&exchange_payload)
+            .map_err(|e| Error::JsonParse(e.to_string()))?;
+        debug!("Sending multi-sig request {res:?}");
+
+        let output = &self
+            .http_client
+            .post("/exchange", res)
+            .await
+            .map_err(|e| Error::JsonParse(e.to_string()))?;
+        debug!("Response: {output}");
+        serde_json::from_str(output).map_err(|e| Error::JsonParse(e.to_string()))
+    }
```
I don't think you are creating this correctly @nicolad 

https://github.com/hyperliquid-dex/hyperliquid-python-sdk/blob/8baad667965968a020a2bb90d0287df0922ca941/hyperliquid/exchange.py#L1068-L1093

v9ai replied:
Thanks, appreciate the reference, will adjust.

v9ai replied:
Please have another look, adjusted.

luca992 replied:
@nicolad I'll actually try it, but looks like it will work if you have access to all the signers which is fine in my case. But this isn't something you can assume in most multisig cases. I think you'd want the function to take Signatures and not PrivateKeySigners? Also might need a way to export a signature. I'd think the usual flow is you collect signatures from the other wallets and when you have enough you as the submitter add your last signature and attach all the other signatures you collected.

This is what the python sdk above appears to do.

```
pub async fn multi_sig_usdc_transfer(
        &self,
        multi_sig_user: Address,
        amount: &str,
        destination: &str,
        wallets: &[PrivateKeySigner],
    )
```

v9ai replied:
Done, thank you.

luca992 reviewed (commented):
I need multi-sign support so I was going over this pr, but I think the final payload is incorrect.

Commit:
Change requests applied

Commit:
Implement proper multi-sig signature collection pattern

luca992 committed:
remove SendAsset payload_multi_sig_user & outer_signer

luca992 commented on src/exchange/actions.rs:
```diff
@@ -210,6 +210,10 @@ pub struct SendAsset {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub payload_multi_sig_user: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub outer_signer: Option<String>,
```
These shouldn't be here. The hyperliquid api returns a deserializaiton an error if you include these

I fixed it in https://github.com/nicolad/hyperliquid-rust-sdk/pull/1

v9ai replied:
Amazing work @luca992, merged.

v9ai replied:
Now hope @lmlmt will get a chance to 👀 over this PR

v9ai committed:
Merge pull request #1 from luca992/multi_sig

remove SendAsset payload_multi_sig_user & outer_signer

luca992 committed:
Revert "remove SendAsset payload_multi_sig_user & outer_signer"

This reverts commit 7b7b4ec33d11d6aaa3b7a695a5972cb527db1b9b.

luca992 committed:
add back outer_signer and payload_multi_sig_user to SendAsset but do not serialize

package outer_signer and payload_multi_sig_user in MultiSigExtension

luca992 committed:
lowercase destinations

luca992 committed:
lowercase payload addresses

luca992 committed:
try signature_chain_id: "0x1"

luca992 committed:
set fixed nonce for debugging

luca992 committed:
log action_without_type

luca992 committed:
serde_json enable preserve_order

luca992 committed:
preserve map order when removing from action

luca992 committed:
insert type at the start of the action

luca992 committed:
Revert "set fixed nonce for debugging"

This reverts commit 9145271bcb62db7ace1cdd344be794ff348c923c.

luca992 committed:
Revert "try signature_chain_id: "0x1""

This reverts commit d1fe9e2049cc14c8798923f11cb5e1ad1942a894.

luca992 committed:
Revert "log action_without_type"

This reverts commit 4451ce2627cfcbfcfe377dfc32864c7a71db0535.

luca992 committed:
Reapply "try signature_chain_id: "0x1""

This reverts commit 3b03853c650f95274daabd8d341028ac6e5ee335.

luca992 committed:
debug with hardcoded nonce

luca992 committed:
Reapply "log action_without_type"

This reverts commit 5d81c388555d1e8b616e5c3bdee4823b06cc9586.

luca992 committed:
log hex

luca992 committed:
Prevent serializing actions to value before needed

This causes issues especially with message packing multi-sig transaction

when for example a serde Value with a number will look like this when packed:

`0x83b07369676e6174757265436861696e4964a3307831aa7369676e6174757265739183a172d942307864623433373135643037383931613063363361383937376139393534646332346664326663363063373436643734353566363133326236396163396665393734a173d942307832386236323363376364643563383230643461663832326634323336623361663361613063373938653736396235353031613331323162343634623761353835a17681bc2473657264655f6a736f6e3a3a707269766174653a3a4e756d626572a23238a77061796c6f616483ac6d756c746953696755736572d92a307832623339633062636664343763356132376636333766383832303661666239356637616661323838ab6f757465725369676e6572d92a307862373131613133636261663734316138383263373230623433363861346566323661346631613632a6616374696f6e8aa474797065a973656e644173736574b07369676e6174757265436861696e4964a730783636656565b068797065726c6971756964436861696ea7546573746e6574ab64657374696e6174696f6ed92a307836643938383538373139363035643332366664633136343839363737333432343462323762386133a9736f75726365446578a0ae64657374696e6174696f6e446578a0a5746f6b656ea455534443a6616d6f756e74a4302e3031ae66726f6d5375624163636f756e74a0a56e6f6e636581bc2473657264655f6a736f6e3a3a707269766174653a3a4e756d626572ad31373632383338313437303030`

```
{
	"signatureChainId": "0x1",
	"signatures": [
		{
			"r": "0xdb43715d07891a0c63a8977a9954dc24fd2fc60c746d7455f6132b69ac9fe974",
			"s": "0x28b623c7cdd5c820d4af822f4236b3af3aa0c798e769b5501a3121b464b7a585",
			"v": {
				"$serde_json::private::Number": "28"
			}
		}
	],
	"payload": {
		"multiSigUser": "0x2b39c0bcfd47c5a27f637f88206afb95f7afa288",
		"outerSigner": "0xb711a13cbaf741a882c720b4368a4ef26a4f1a62",
		"action": {
			"type": "sendAsset",
			"signatureChainId": "0x66eee",
			"hyperliquidChain": "Testnet",
			"destination": "0x6d98858719605d326fdc1648967734244b27b8a3",
			"sourceDex": "",
			"destinationDex": "",
			"token": "USDC",
			"amount": "0.01",
			"fromSubAccount": "",
			"nonce": {
				"$serde_json::private::Number": "1762838147000"
			}
		}
	}
}
```

luca992 committed:
Revert "debug with hardcoded nonce"

This reverts commit b79d86ad567c78b8847f31bff5150498a348e4e8.

luca992 committed:
serialize Signature r & s to strings

luca992 committed:
Reapply "debug with hardcoded nonce"

This reverts commit c660e12a5b57dfdb1eab7586db605fa4df8506f6.

luca992 committed:
MultiSigEnvelope use the signature_chain_id from the MultiSigAction

luca992 committed:
Revert "Reapply "debug with hardcoded nonce""

This reverts commit 700679d3879d37fff366b3a6e3f150fe33f303b5.

luca992 committed:
revert debug print

luca992 committed:
revert adding preserve_order serde_json feature

luca992 committed:
Revert "revert adding preserve_order serde_json feature"

This reverts commit 5ef467f6f8c301b131fbf09c9a299412ee228a4d.

luca992 commented on src/signature/create_signature.rs:
```diff
@@ -27,6 +27,197 @@ pub(crate) fn sign_typed_data<T: Eip712>(
+
+    let mut bytes = rmp_serde::to_vec_named(&action_without_type)
+        .map_err(|e| Error::RmpParse(e.to_string()))?;
+
```
https://github.com/nicolad/hyperliquid-rust-sdk/pull/2

Got it working, but had to make a lot of fixes.

Also had to add back `payload_multi_sig_user` & `outer_signer` but with serialization disabled

v9ai replied:
Hey @luca992, isn't this the beauty of OS contribution? Thx.

luca992 reviewed (changes requested):
https://github.com/nicolad/hyperliquid-rust-sdk/pull/2

Side note the multisig apis you made are using `SendAsset` not `UsdSend`... it works (You can use either to send usd) but is not really named correctly.

v9ai committed:
Merge pull request #2 from luca992/multi_sig

Multi sig fixes

v9ai commented:
@lmlmt 👋 
if you have chance, could you review this?

lmlmt commented:
> @lmlmt 👋 if you have chance, could you review this?

Hi, please circulate pull requests against https://github.com/infinitefield/hypersdk going forward. This repo is not actively maintained anymore.

v9ai commented:
> > @lmlmt 👋 if you have chance, could you review this?
> 
> Hi, please circulate pull requests against https://github.com/infinitefield/hypersdk going forward. This repo is not actively maintained anymore.

@lmlmt having this info in the README would’ve helped a lot

luca992 commented:
@lmlmt ... Is that a fork or a new implementation? If it's a a whole new code base I'll probably just fork this repo and republish it as I have no problems with this multi-sig implementation
